### ISSUE TITLE
TranslationWalker does not run when Query Cache is enabled

### ISSUE THREAD

**Issue by flip111** [doctrine-extensions/DoctrineExtensions#1021]
I had a bug with translatable, it was firing queries even though registering the walker and hinting the right locale was done properly. After tracing the problem i came to the conclusion that Query Caching should be disabled for queries using Translatable.
To disable the query cache on a query: `$query->useQueryCache(false);`

Reasoning:
If query cache is used the Parser is never invoked:
https://github.com/doctrine/doctrine2/blob/5c44dd38234f2e143e376ac94d2e7f945d7f9328/lib/Doctrine/ORM/Query.php#L248

If the Parser does not run then no custom output walkers are checked:
https://github.com/doctrine/doctrine2/blob/4f585a3d6372ef1115a30663fe0005f611decd76/lib/Doctrine/ORM/Query/Parser.php#L357

I think the difference between a `HINT_CUSTOM_TREE_WALKERS` (SQL Walker) and `HINT_CUSTOM_OUTPUT_WALKER` (output walker) is that the first changes the actual DQL query and the latter always has to be done afterwards.

This is possibly a limitation of Doctrine

To do: gedmo translatable should automatically turn off query caching when it needs to register an output walker

**Comment by acasademont** [comment]
Yep, same here. Even though the query is executed correctly with the translation joins the "skipOnLoad" method is not called and the translation objects are fetched...

**Comment by l3pp4rd** [comment]
hints are used in cache key, it also is mentioned in the documentation. should be clear, read carefully

**Comment by acasademont** [comment]
@l3pp4rd I have already read the docs,  and I have even contributed extensively in the past to the Translatable component and I perfectly know how it works, no need to be rude here.

It has nothing to do with hints. When the query is cached the custom translatable hydrators are not executed (these hydrators are set on the `TranslationWalker` and the walker is obviously never run when the query is already cached) and therefore the `setSkipOnLoad` method is not called and the `TranslatableListener` does again all the work.

The workaround is to manually set the custom translatable hydrators when executing the query.

So I would kindly ask you to reopen the issue.

**Comment by flip111** [comment]
@l3pp4rd 
I believe only the original query is cached (inside the querycache, not the resultcache). The post-processing done by hints on the query yields a new query ... and so i think this new query (with join for translate table) is NOT cached in the query cache. If this is not true `hints are used in cache key` please be more explicit to the actual code where the modified query is cached.

@acasademont i don't understand your workaround, what do you mean with `custom translatable hydrator` ?

**Comment by acasademont** [comment]
When the query is run without caching, the `TranslatableWalker` fires and this walker is responsible for appending all the necessary joins. Besides, it also changes the default ObjectHydrator by a custom ObjectHydrator. This custom ObjectHydrator the only thing that does is  a call to `setSkipOnLoad` on the `TranslatableListener`.

When the `postLoad` event in the `TranslatableListener` kicks in it checks for this `skipOnLoad` property. If the property is not set, it will run the normal behaviour of querying the translation table to translate the entity. When we already have the needed information in the query results thanks to the walker sql statements there's no need to perform these actions, that's where the performance improves A LOT.

BUT, if the query is cached, the `TranslationWalker` is not run, and the custom ObjectHydrators are not set and the `skipOnLoad` property is not set and we run the queries again. This is a performance disaster as we are already joining the translation tables AND fetching this info again.

**Comment by flip111** [comment]
@acasademont ah yes, i remember now (given i made this issue 1.5 years ago). I think your workaround is better than mine since you can still use the querycache. Perhaps a nice idea if you could write a small snippet how to set the custom hydrator, that way other people can benefit (i'm not working with Doctrine anymore myself).

And about performance ... yes this is the major issue. The amount of queries this generate is huge and definitively noticeable in normal use.

**Comment by l3pp4rd** [comment]
@acasademont hm misunderstood the issue. probably we need a better way to deal with skipOnLoad behavior if that is a cause in general to reload the field translations. as far as I remember the older versions of ORM were not causing this problem.

**Comment by pmishev** [comment]
Are there any new ideas regarding this issue?
In certain scenarios the workaround proposed by @acasademont for setting custom hydrator when you execute the query is not feasible, because you have no control over the query. An example is when using Symfony forms:
```
        $builder->add('sectorLinks', EntityType::class, [
            'class'         => 'AppBundle:Sector',
            'query_builder' => function (EntityRepository $er) {
                return $er->createQueryBuilder('s')->orderBy('s.left', 'ASC');
            }
```
When executing the query, the queryCache is hit and the custom object hydrator is not set.

I could set globally `query_cache_driver: void`, which would solve this problem, but turning off the query cache would have performance implications on its own, so I don't consider it a solution.

**Comment by ConneXNL** [comment]
Just stumbled across this problem using Doctrine ORM version v2.5.4 and DoctrineExtensions v2.4.1.
Sadly took me 4h to find out I had to add this one liner :
`$query->useQueryCache(false);`

**Comment by pitpit** [comment]
Same problem here , using following hints on query
```
        $query->setHint(Query::HINT_CUSTOM_OUTPUT_WALKER, TranslationWalker::class);
        $query->setHint(TranslatableListener::HINT_TRANSLATABLE_LOCALE, $locale);
        $query->setHint(TranslatableListener::HINT_INNER_JOIN, true);
```

vendor versions : doctrine/orm: 2.5.10, gedmo/doctrine-extensions:v2.4.30, stof/doctrine-extensions-bundle:v1.2.2

Using `$query->useQueryCache(false);` or `doctrine.orm.query_cache_driver` set to `array` is a workaround but not acceptable for performance considerations

**Comment by pitpit** [comment]
An update to my previous post, Karol Wojciechowski's answer to this [stackoverflow](https://stackoverflow.com/a/27173280) solves my issue :

```php
$query->setHint(Query::HINT_CUSTOM_OUTPUT_WALKER, TranslationWalker::class);
$query->setHint(TranslatableListener::HINT_TRANSLATABLE_LOCALE, $locale);
$query->setHint(TranslatableListener::HINT_INNER_JOIN, true);

$config = $this->getEntityManager()->getConfiguration();
if ($config->getCustomHydrationMode(TranslationWalker::HYDRATE_OBJECT_TRANSLATION) === null) {
    $config->addCustomHydrationMode(
        TranslationWalker::HYDRATE_OBJECT_TRANSLATION,
        ObjectHydrator::class
    );
}

return $query->getResult(TranslationWalker::HYDRATE_OBJECT_TRANSLATION);
```

**Comment by pitpit** [comment]
... but TranslationWalker::HYDRATE_OBJECT_TRANSLATION does not hydrate associated entities....

**Comment by piepke2000** [comment]
I have the same issue. It keeps querying for the translations, although using the TranslationWalker. I tried about everything I can find. 
I use doctrine/orm version 2.4.7 and gedmo/doctrine-extensions version 2.3.12

My code below:

`$q = "SELECT p FROM MyBundle:Product p WHERE p.isActive = 1";

        $query = $this->getEntityManager()->createQuery($q);

        $query->useQueryCache(false);

        $query->setHint(Query::HINT_CUSTOM_OUTPUT_WALKER, TranslationWalker::class);
        $query->setHint(TranslatableListener::HINT_TRANSLATABLE_LOCALE, 'nl');
        $query->setHint(TranslatableListener::HINT_INNER_JOIN, true);

        $config = $this->getEntityManager()->getConfiguration();
        if ($config->getCustomHydrationMode(TranslationWalker::HYDRATE_OBJECT_TRANSLATION) === null) {
            $config->addCustomHydrationMode(
                TranslationWalker::HYDRATE_OBJECT_TRANSLATION,
                ObjectHydrator::class
            );
        }

        $res = $query->getResult(TranslationWalker::HYDRATE_OBJECT_TRANSLATION);`

When I use getArrayResult() it works as expected. 
Anybody with an idea? I was thinking it had something to do with the versions, but I can't find anything on that.

**Comment by pitpit** [comment]
4 years later, I face this bug once again...

* gedmo/doctrine-extensions:3.0.5
* doctrine/dbal:2.13.1

![image](https://user-images.githubusercontent.com/283481/117146976-10f35380-adb5-11eb-92da-17a476428634.png)

**Comment by stollr** [comment]
I faced this issue, today. Debugging it and finally finding the reason and this issue is really painful. But this way I found another workaround to address the issue.

The `TranslationWalker` adds the `Query::HINT_REFRESH` to the query which leads to an update of the entity with the correct locale, event if it was already loaded with another locale before.

But the hint is not cached, so the update of the entity does not happen anymore once the cached query is used.

But one can manually add the `Query::HINT_REFRESH` to the query, so it even works if the query is loaded from the cache. I think this should be added to the documentation, right?
